### ISSUE TITLE
General Re-working of HTTPS Info Extension

### ISSUE THREAD

**Issue by zapbot** [zaproxy/zaproxy#758]
```
Enhancement request for the HTTPInfo add-on beast/crime tests to also be converted into
an active scan rule.


```

Original issue reported on code.google.com by `psiinon` on 2013-08-06 15:11:47

**Comment by kingthorin** [comment]
The author of the original code base has moved to C#, and makes arguments against testing BEAST:
http://www.bolet.org/TestSSLServer/

I'll reach out to see if I can get the latest in his java code base.

Changing these to active scanners would mean implementing much of the required functionality as an active scanner. Meaning that protocol support and cipher suite enumeration would have to happen per host, which is likely to exceed the requests/strength guidelines for the majority of targets.

I'd like to suggest that we don't move them, but do add raising alerts from within the httpsinfo addon. Does that make sense? If so I'll start investigating that angle. (Which may be enhanced by an updated java code base from the original author but is not dependent upon it...)

**Comment by kingthorin** [comment]
Got a reply from the original dev, he'd already moved to C# when the last Java version was published so there's no update to be had.

**Comment by thc202** [comment]
Yes, if the user controls when and what alerts are raised.

**Comment by kingthorin** [comment]
Could you be more specific?

The user controls the "scanning" (the when) by using the addon. By controlling "what" do you mean they should have control over whether or not alerts are raised by httpdinfo?

Or maybe you envision something simple and I'm over complicating? Or perhaps you envision greater control or re-engineering and I'm making it too simple?

**Comment by thc202** [comment]
Sure. About the "when", from the name of the menu item ("Show HTTPS information") I'd not expect having them automatically raised, maybe through a button? But I guess it could also raise the alerts automatically, it's just a matter of rewording the menu item to use "scan" instead.
For the "what", there are couple of alerts that could be raised from the info gathered (encryption strengths, CRIME, BEAST).

**Comment by kingthorin** [comment]
Ok, I'll investigate further and come up with some ideas.

Any thoughts on the argument against testing/reporting BEAST?

**Comment by thc202** [comment]
We could still allow that, but in the hypothetical issue mention the arguments/link in solution/refs.
That could also be added to the help pages.

**Comment by kingthorin** [comment]
Ok, that makes sense.

**Comment by kingthorin** [comment]
Any objection to me renaming this issue? Something like "General Re-working of HTTPS Info Extension". 

The plan: Aligning it to the current SimpleExample. Allowing isSafe() to return false, making the menu item name more clear (implying scanning for HTTPS Issues), allowing it to raise Alerts for Beast/Crime (whatever else might be added down the road) [perhaps leveraging alertFingerPrint to prevent duplicate entries], perhaps updating or replacing SSLServer.java (which as discussed in earlier comments is baed on 2012 code which is no longer maintained or updated).

**Comment by thc202** [comment]
Sounds good to me.

**Comment by psiinon** [comment]
+1

**Comment by kingthorin** [comment]
TODO List:

- [ ]  Format code (see: https://github.com/zaproxy/zap-extensions/issues/349)
- [ ] Internationalize current interface Strings
- [ ] Align with SimpleExample (implement isSafe(), etc) (see https://github.com/zaproxy/zap-extensions/pull/343)
- [ ] Beast/Crime checks raise alerts 
- [ ] Prevent alerts from being duplicated by checking alertFingerPrint() ?
- [ ] Change from dialog to tab
- [ ] Get SSL (Lock) icon (if we don't already have one) [related to tab setup]
- [ ] Look at updates to or replacement of SSLServer.java
- [ ] Update help
- [ ] Evaluate adding additional SSL/TLS related checks (and alerts)

**Comment by thc202** [comment]
By default the menu items are not safe, in this case it shouldn't require any further changes.

Suggestions to TODO list:
- Rename the extension class (to ExtensionHttpsInfo?);
- Change package name to lower case.

**Comment by thc202** [comment]
@kingthorin about review comments, should those be added to the linked commits? Or, better wait for a pull request? Or?

**Comment by kingthorin** [comment]
I was just trying to keep track of things. If you have a look at any of it and something really stands out then let me know I guess, but I'm fine waiting for a PR.

**Comment by thc202** [comment]
OK, I'll wait for a pull request then :)

**Comment by kingthorin** [comment]
Need opinions on an icon for this extension: http://p.yusukekamiyamane.com/icons/search/fugue/#keyword=lock

I think either the lock icon with the SSL label or the lock icon with the blue arrow.

**Comment by thc202** [comment]
I'd go with the icon with SSL label, it gives more information/context.

**Comment by kingthorin** [comment]
I think at this point the one remaining optional item and all future items should be split off into another issue. None of them are simple or straight forward.

A significant amount of progress has been made which should probably be released to the community.

Agreed?

**Comment by thc202** [comment]
Sure.

**Comment by psiinon** [comment]
Sounds good

**Comment by thc202** [comment]
Fixed in zaproxy/zap-extensions#369.
